File: app/src/main/java/com/mdpgrp4/mdpremote/ControllerActivity.java
package com.mdpgrp4.mdpremote;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.IBinder;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

public class ControllerActivity extends AppCompatActivity {


    //Shared Preferences
    public static final String DEFAULT = "N/A";
    BluetoothService mService;
    boolean mBound = false;
    String functionA, functionB;
    //Buttons
    private Button functionABtn;
    private Button functionBBtn;
    private Button forwardBtn;
    private Button reverseBtn;
    private Button leftBtn;
    private Button rightBtn;
    private Button functionBSave;
    private Button functionASave;
    //EditText
    private EditText functionAEdit;
    private EditText functionBEdit;
    private ServiceConnection mConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            BluetoothService.BluetoothBinder binder = (BluetoothService.BluetoothBinder) iBinder;
            mService = binder.getService();
            mBound = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            mBound = false;
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_controller);

        functionASave = (Button) findViewById(R.id.functionASave);
        functionBSave = (Button) findViewById(R.id.functionBSave);
        functionABtn = (Button) findViewById(R.id.functionABtn);
        functionBBtn = (Button) findViewById(R.id.functionBBtn);
        forwardBtn = (Button) findViewById(R.id.forwardBtn);
        reverseBtn = (Button) findViewById(R.id.reverseBtn);
        leftBtn = (Button) findViewById(R.id.leftBtn);
        rightBtn = (Button) findViewById(R.id.rightBtn);

        functionAEdit = (EditText) findViewById(R.id.functionAEdit);
        functionBEdit = (EditText) findViewById(R.id.functionBEdit);

        functionASave.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                saveFunctionData(functionAEdit, "function A");
            }
        });
        functionBSave.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                saveFunctionData(functionBEdit, "function B");
            }
        });

        loadFunctionData();

        functionABtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mBound) {
                    mService.writeBtOut(functionAEdit.getText().toString());
                }
            }
        });
        functionBBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mBound) {
                    mService.writeBtOut(functionBEdit.getText().toString());
                }
            }
        });
        forwardBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mBound) {
                    mService.writeBtOut("f");
                }
            }
        });
        reverseBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mBound) {
                    mService.writeBtOut("r");
                }
            }
        });
        leftBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mBound) {
                    mService.writeBtOut("tl");
                }
            }
        });
        rightBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mBound) {
                    mService.writeBtOut("tr");
                }
            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);
        Intent bluetoothIntent = new Intent(this, BluetoothService.class);
        bindService(bluetoothIntent, mConnection, BIND_ABOVE_CLIENT);
    }

    @Override
    protected void onStop() {
        EventBus.getDefault().unregister(this);
        if (mBound) {
            unbindService(mConnection);
            mBound = false;
        }
        super.onStop();
    }

    public void saveFunctionData(EditText text, String name) {

        SharedPreferences sp = getSharedPreferences("MyData", Context.MODE_PRIVATE);
        SharedPreferences.Editor editSP = sp.edit();

        editSP.putString(name, text.getText().toString());

        editSP.apply();

        Toast.makeText(this, "Saved " + name + " successfully", Toast.LENGTH_SHORT).show();
    }

    //to load shared preferences
    public void loadFunctionData() {
        SharedPreferences sp = getSharedPreferences("MyData", Context.MODE_PRIVATE);
        functionA = sp.getString("function A", DEFAULT);
        functionB = sp.getString("function B", DEFAULT);
        if(functionA.equals(DEFAULT) && functionB.equals(DEFAULT)){
            Toast.makeText(this, "No functions data found", Toast.LENGTH_SHORT).show();
            functionAEdit.setText(DEFAULT);
            functionBEdit.setText(DEFAULT);
        }else{
            Toast.makeText(this, "Functions loaded", Toast.LENGTH_SHORT).show();
            functionAEdit.setText(functionA);
            functionBEdit.setText(functionB);
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onBluetoothEvent(MessageEvent event) {
        switch (event.status) {
            case MessageEvent.DISCONNECTED:
                finish();
                break;
            case MessageEvent.ROBOT_STATUS:
                Toast.makeText(this, event.message, Toast.LENGTH_SHORT).show();
                break;
        }
    }
}

File: app/src/main/java/com/mdpgrp4/mdpremote/ControllerActivity.kt
package com.mdpgrp4.mdpremote

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class ControllerActivity : AppCompatActivity() {
    internal var mService: BluetoothService? = null
    internal var mBound = false
    internal var functionA: String? = null
    internal var functionB: String? = null
    //Buttons
    private var functionABtn: Button? = null
    private var functionBBtn: Button? = null
    private var forwardBtn: Button? = null
    private var reverseBtn: Button? = null
    private var leftBtn: Button? = null
    private var rightBtn: Button? = null
    private var functionBSave: Button? = null
    private var functionASave: Button? = null
    //EditText
    private var functionAEdit: EditText? = null
    private var functionBEdit: EditText? = null

    private var statusView: TextView? = null

    private val mConnection = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            val binder = iBinder as BluetoothService.BluetoothBinder
            mService = binder.service
            mBound = true
        }

        override fun onServiceDisconnected(componentName: ComponentName) {
            mBound = false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_controller)

        statusView = findViewById<View>(R.id.statusTextView) as TextView

        functionASave = findViewById<View>(R.id.functionASave) as Button
        functionBSave = findViewById<View>(R.id.functionBSave) as Button
        functionABtn = findViewById<View>(R.id.functionABtn) as Button
        functionBBtn = findViewById<View>(R.id.functionBBtn) as Button
        forwardBtn = findViewById<View>(R.id.forwardBtn) as Button
        reverseBtn = findViewById<View>(R.id.reverseBtn) as Button
        leftBtn = findViewById<View>(R.id.leftBtn) as Button
        rightBtn = findViewById<View>(R.id.rightBtn) as Button

        functionAEdit = findViewById<View>(R.id.functionAEdit) as EditText
        functionBEdit = findViewById<View>(R.id.functionBEdit) as EditText

        functionASave!!.setOnClickListener { saveFunctionData(functionAEdit, "function A") }
        functionBSave!!.setOnClickListener { saveFunctionData(functionBEdit, "function B") }

        loadFunctionData()

        functionABtn!!.setOnClickListener {
            if (mBound) {
                mService!!.writeBtOut(functionAEdit!!.text.toString())
            }
        }
        functionBBtn!!.setOnClickListener {
            if (mBound) {
                mService!!.writeBtOut(functionBEdit!!.text.toString())
            }
        }
        forwardBtn!!.setOnClickListener {
            if (mBound) {
                mService!!.writeBtOut("sr17")
                statusView!!.text = "Moving forward"
            }
        }
        reverseBtn!!.setOnClickListener {
            if (mBound) {
                mService!!.writeBtOut("sr64")
                statusView!!.text = "Reversing"
            }
        }
        leftBtn!!.setOnClickListener {
            if (mBound) {
                mService!!.writeBtOut("sr32")
                statusView!!.text = "Turning left"
            }
        }
        rightBtn!!.setOnClickListener {
            if (mBound) {
                mService!!.writeBtOut("sr48")
                statusView!!.text = "Turning right"
            }
        }
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
        val bluetoothIntent = Intent(this, BluetoothService::class.java)
        bindService(bluetoothIntent, mConnection, Context.BIND_ABOVE_CLIENT)
    }

    override fun onStop() {
        EventBus.getDefault().unregister(this)
        if (mBound) {
            unbindService(mConnection)
            mBound = false
        }
        super.onStop()
    }

    fun saveFunctionData(text: EditText?, name: String) {

        val sp = getSharedPreferences("MyData", Context.MODE_PRIVATE)
        val editSP = sp.edit()

        editSP.putString(name, text!!.text.toString())

        editSP.apply()

        Toast.makeText(this, "Saved $name successfully", Toast.LENGTH_SHORT).show()
    }

    //to load shared preferences
    fun loadFunctionData() {
        val sp = getSharedPreferences("MyData", Context.MODE_PRIVATE)
        functionA = sp.getString("function A", DEFAULT)
        functionB = sp.getString("function B", DEFAULT)
        if (functionA == DEFAULT && functionB == DEFAULT) {
            Toast.makeText(this, "No functions data found", Toast.LENGTH_SHORT).show()
            functionAEdit!!.setText(DEFAULT)
            functionBEdit!!.setText(DEFAULT)
        } else {
            Toast.makeText(this, "Functions loaded", Toast.LENGTH_SHORT).show()
            functionAEdit!!.setText(functionA)
            functionBEdit!!.setText(functionB)
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onBluetoothEvent(event: MessageEvent) {
        when (event.status) {
            MessageEvent.DISCONNECTED -> finish()
            MessageEvent.ROBOT_STATUS -> statusView!!.text = event.message
        }
    }

    companion object {


        //Shared Preferences
        val DEFAULT = "N/A"
    }
}
